File: bcs-services/bcs-helm-manager/internal/store/entity/repo.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entity

import (
	"context"
	"time"

	"github.com/Tencent/bk-bcs/bcs-services/bcs-helm-manager/internal/common"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-helm-manager/internal/i18n"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-helm-manager/internal/options"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-helm-manager/internal/repo"
	helmmanager "github.com/Tencent/bk-bcs/bcs-services/bcs-helm-manager/proto/bcs-helm-manager"
)

// Repository 定义了仓库基础信息, 存储在helm-manager的数据库中, 是chart相关操作的基础
type Repository struct {
	// basic info
	TenantID          string `json:"tenantID" bson:"tenantID"`
	TenantProjectCode string `json:"tenantProjectCode" bson:"tenantProjectCode"`
	ProjectID         string `json:"projectID" bson:"projectID"`
	Name              string `json:"name" bson:"name"`
	DisplayName       string `json:"displayName" bson:"displayName"`
	Personal          bool   `json:"personal" bson:"personal"`
	Public            bool   `json:"public" bson:"public"`
	Type              string `json:"type" bson:"type"`
	RepoURL           string `json:"repoURL" bson:"repoURL"`

	// remote repo settings
	Remote         bool   `json:"remote" bson:"remote"`
	RemoteURL      string `json:"remoteURL" bson:"remoteURL"`
	RemoteUsername string `json:"remoteUsername" bson:"remoteUsername"`
	RemotePassword string `json:"remotePassword" bson:"remotePassword"`

	// auth
	Username string `json:"username" bson:"username"`
	Password string `json:"password" bson:"password"`

	CreateBy   string `json:"createBy" bson:"createBy"`
	UpdateBy   string `json:"updateBy" bson:"updateBy"`
	CreateTime int64  `json:"createTime" bson:"createTime"`
	UpdateTime int64  `json:"updateTime" bson:"updateTime"`
}

// GetRepoProjectID get repo project id
func (r *Repository) GetRepoProjectID() string {
	if r.Public {
		return repo.GetBKRepoProjectID("system", options.GlobalOptions.Repo.PublicRepoProject)
	}
	projectCode := r.ProjectID
	if r.TenantProjectCode != "" {
		projectCode = r.TenantProjectCode
	}
	return repo.GetBKRepoProjectID(r.TenantID, projectCode)
}

// GetRepoName get repo name
func (r *Repository) GetRepoName() string {
	if r.Public {
		return options.GlobalOptions.Repo.PublicRepoName
	}
	return r.Name
}

// Transfer2Proto transfer the data into protobuf struct
func (r *Repository) Transfer2Proto(ctx context.Context) *helmmanager.Repository {
	displayName := r.DisplayName
	if len(displayName) == 0 {
		displayName = common.ProjectRepoDefaultDisplayName
	}

	// 语言转换
	displayName = i18n.GetMsg(ctx, displayName)

	return &helmmanager.Repository{
		ProjectCode: common.GetStringP(r.ProjectID),
		Name:        common.GetStringP(r.Name),
		DisplayName: common.GetStringP(displayName),
		Public:      common.GetBoolP(r.Public),
		Type:        common.GetStringP(r.Type),
		RepoURL:     common.GetStringP(r.RepoURL),
		Remote:      common.GetBoolP(r.Remote),
		RemoteURL:   common.GetStringP(r.RemoteURL),
		Username:    common.GetStringP(r.Username),
		Password:    common.GetStringP(r.Password),
		CreateBy:    common.GetStringP(r.CreateBy),
		UpdateBy:    common.GetStringP(r.UpdateBy),
		CreateTime:  common.GetStringP(time.Unix(r.CreateTime, 0).UTC().Format(time.RFC3339)),
		UpdateTime:  common.GetStringP(time.Unix(r.UpdateTime, 0).UTC().Format(time.RFC3339)),
	}
}

// LoadFromProto load data from protobuf struct
func (r *Repository) LoadFromProto(repository *helmmanager.Repository) M {
	if repository == nil {
		return nil
	}
	m := make(M)

	if repository.ProjectCode != nil {
		r.ProjectID = repository.GetProjectCode()
		m[FieldKeyProjectID] = r.ProjectID
	}
	if repository.Name != nil {
		r.Name = repository.GetName()
		m[FieldKeyName] = r.Name
	}
	if repository.DisplayName != nil {
		r.DisplayName = repository.GetDisplayName()
		m[FieldKeyDisplayName] = r.Name
	}
	if repository.Public != nil {
		r.Public = repository.GetPublic()
		m[FieldKeyPublic] = r.Name
	}
	if repository.Type != nil {
		r.Type = repository.GetType()
		m[FieldKeyType] = r.Type
	}
	if repository.Remote != nil {
		r.Remote = repository.GetRemote()
		m[FieldKeyRemote] = r.Remote
	}
	if repository.RemoteURL != nil {
		r.RemoteURL = repository.GetRemoteURL()
		m[FieldKeyRemoteURL] = r.RemoteURL
	}
	if repository.RemoteUsername != nil {
		r.RemoteUsername = repository.GetRemoteUsername()
		m[FieldKeyRemoteUsername] = r.RemoteUsername
	}
	if repository.RemotePassword != nil {
		r.RemotePassword = repository.GetRemotePassword()
		m[FieldKeyRemotePassword] = r.RemotePassword
	}
	if repository.RepoURL != nil {
		r.RepoURL = repository.GetRepoURL()
		m[FieldKeyRepoURL] = r.RepoURL
	}
	if repository.Username != nil {
		r.Username = repository.GetUsername()
		m[FieldKeyUsername] = r.Username
	}
	if repository.Password != nil {
		r.Password = repository.GetPassword()
		m[FieldKeyPassword] = r.Password
	}
	if repository.CreateBy != nil {
		r.CreateBy = repository.GetCreateBy()
		m[FieldKeyCreateBy] = r.CreateBy
	}
	if repository.UpdateBy != nil {
		r.UpdateBy = repository.GetUpdateBy()
		m[FieldKeyUpdateBy] = r.UpdateBy
	}

	return m
}

File: bcs-ui/backend/helm/repository/repo.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
import logging
from typing import Dict, Tuple

from django.conf import settings

from backend.components.bk_repo import BkRepoClient, BkRepoTokenError
from backend.helm.helm.models.repo import Repository
from backend.helm.repository.constants import RepoCategory
from backend.utils.basic import getitems
from backend.utils.error_codes import error_codes

logger = logging.getLogger(__name__)


def get_repo_addr(project_code: str, repo_name: str) -> str:
    """获取仓库地址"""
    # 仓库地址: helm-repo-domain/project_code/repo_name
    return f"{settings.HELM_REPO_DOMAIN}/{project_code}/{repo_name}"


def get_repo_auth(username: str, project_code: str) -> Tuple[str, str]:
    """获取仓库的授权信息，包含 username 和 password"""
    client = BkRepoClient(username=username)
    # 获取用户对应的 token
    token = {}
    try:
        token = client.get_token()
    except BkRepoTokenError as e:
        logger.exception("获取token失败, %s", e)
    # 如果可以获取到token，则直接返回
    if token:
        return (username, token["id"])
    # 生成token
    token = client.set_token(project_code)
    return (username, token["id"])


def get_repo(access_token: str, username: str, project_code: str, repo_name: str) -> Dict:
    """获取单个repo"""
    # 通过接口列表匹配单个仓库，并返回仓库信息
    # 获取仓库列表
    repos = BkRepoClient(access_token=access_token, username=username).list_project_repos(project_code)
    # 匹配仓库，并判断仓库是否为纳管仓库
    repo_info = {}
    for repo in repos:
        if repo["name"] != repo_name:
            continue
        repo_info = repo
    # 没有查到是抛出异常
    if not repo_info:
        raise error_codes.ResNotFoundError(f"repo: {repo_name} not found")
    repo_info["is_imported"] = is_imported(repo_info)
    repo_info["addr"] = get_repo_addr(project_code, repo_info["name"])
    # 添加访问制品库的用户名和token
    repo_info["username"], repo_info["password"] = get_repo_auth(username, project_code)

    return repo_info


def is_imported(repo: Dict) -> bool:
    if repo["category"] == RepoCategory.REMOTE:
        return True
    if repo["category"] == RepoCategory.COMPOSITE and getitems(repo, ["configuration", "proxy", "channelList"]):
        return True
    return False


def is_imported_repo(access_token: str, username: str, project_code: str, repo_name: str) -> bool:
    """判断是否为纳管的仓库

    NOTE: 这里通过仓库列表进行匹配，没有单个仓库接口
    """
    repo = get_repo(access_token, username, project_code, repo_name)
    # 判断条件: category为COMPOSITE和REMOTE，当为COMPOSITE类型时，channelList(代理源)不为空
    return is_imported(repo)


class RepoDBActions:
    def __init__(self, project_id: str, project_code: str, repo_name: str):
        self.project_id = project_id
        self.project_code = project_code
        self.repo_name = repo_name

    def get_or_create(self) -> Repository:
        # url 的组装格式: helm-repo-domain/project_code/
        repo_params = {
            "url": get_repo_addr(self.project_code, self.repo_name),
            "provider": "bkrepo",
            "storage_info": {},
        }
        return Repository.objects.get_or_create(
            name=self.repo_name,
            project_id=self.project_id,
            defaults=repo_params,
        )[0]

    def delete(self) -> None:
        """删除 repo"""
        Repository.objects.filter(project_id=self.project_id, name=self.repo_name).delete()
        return
